tensorflow/core/graph/mkl_optimizer_merge.h

```diff
@@ -0,0 +1,217 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+// An optimization pass that performs node merging and rewrite on graph nodes
+
+#ifndef TENSORFLOW_CORE_GRAPH_MKL_OPTIMIZER_MERGE_H_
+#define TENSORFLOW_CORE_GRAPH_MKL_OPTIMIZER_MERGE_H_
+
+#ifdef INTEL_MKL
+
+#include <sys/types.h>
+#include <vector>
+#include <string>
+#include <memory>
+#include "tensorflow/core/graph/graph.h"
+#include "tensorflow/core/lib/core/status.h"
+#include "tensorflow/core/common_runtime/optimization_registry.h"
+
+namespace tensorflow {
+
+// How many hops do we search for matching node in the backward dataflow graph?
+// We use maxhop of 10 based on empirical observations. Also, these are
+// maxhops in backward data-flow graph. Since input of forward nodes (Conv2D)
+// directly goes to backward nodes, we do not expect the hop-distance
+// would be more than few nodes.
+#define NODEMERGE_CONTEXT_MAXDEPTH 10
+
+// This optimization pass performs two tasks: merge
+// nodes in the forward pass, and rewrite the gradient ops
+// corresponding to merged forward ops.
+//
+// Merging nodes in the graph: Currently, it merges Conv2D+AddBias
+// and MatMul+AddBias nodes together.
+//
+// Rewriting nodes in the graph:
+// This is neded in order to optimize gradient ops of
+// Conv2D+AddBias and MatMul+AddBias. Gradient op of
+// both the Conv2D and MatMul is BiasAddGrad, and we
+// need to rewrite BiasAddGrad into Conv2D-specific BiasAddGrad,
+// and MatMul-specific BiasAddGrad. This is context-specific
+// optimization, where the context is the forward operator
+// that the BiasAddGrad corresponds to.
+//
+class NodeMergeRewritePass : public GraphOptimizationPass {
+ public:
+  NodeMergeRewritePass() {
+    csinfo_.conv2d                     = "Conv2D";
+    csinfo_.conv2dwithbias             = "Conv2DWithBias";
+    csinfo_.conv2dwithbiasbackpropbias = "Conv2DWithBiasBackpropBias";
+    csinfo_.biasadd                    = "BiasAdd";
+    csinfo_.matmul                     = "MatMul";
+    csinfo_.matmulmkl                  = "MatMulMkl";
+    csinfo_.add                        = "Add";
+    csinfo_.biasaddgrad                = "BiasAddGrad";
+
+    minfo_.push_back({csinfo_.conv2d, csinfo_.biasadd, 0,
+                      csinfo_.conv2dwithbias});
+    // NOTE: We no longer use this optimization now.
+    // minfo_.push_back({csinfo_.matmul, csinfo_.add, 0,
+    //                csinfo_.matmulmkl});
+
+    // We use maxhop of 10 based on emperical observations. Also, these are
```

Comment:
Ok, so its not a correctness issue then, correct? i.e. it will always terminate, it just might take longer to terminate if you did not use maxhop?

I would re-work your comment about maxhop to start with the description, then add your comment about "we chose a maxhop of 10 based on emperical observations".